### PR TITLE
Extract grading functionality from `Grader` to `Applications`

### PR DESCRIPTION
**Pull request by pberkes**
Grader is still not much covered by test, so please give it a spin.

### CODE REVIEW

**Comment by pberkes** [comment]
I assumed that the answer to #12 is "yes"

**Comment by pberkes** [comment]
This PR cleans up a little bit and adds a couple of tests. Is it interesting at all? If so I'll resolve the conflicts, but then you need to promise to review it!

**Comment by jni** [comment]
I'll review it!

**Comment by pberkes** [comment]
BTW this PR should solve the problem of "identities"

**Comment by pberkes** [comment]
@jni I merged `master` in this branch and solved the conflicts. The tests pass.

I cannot test it on real application: I have access to the repository, but even `master` fails to start with those... Could you please have a look?

**Comment by otizonaizit** [comment]
This is so painful, man. In order to get it to work you need two things:

  - on the pythonschool repo, checkout commit `ebad5eb` (this will give you the state after the ratings had been done and before the cleanup after the end of the Camerino school)
  - on grader, from master I had to go back to `89792b0` to have something that works
  - change to `pythonschool/applications` and launch `grader`: it should work now

My take on this:
 - after `89792b0` there have been many commits by @keszybz : no idea what he did exactly, but they break current grader
 - to avoid the dance of having to checkout an old commit to be able to work with old applications, the whole `grader.conf`+`applications.csv` loading logic needs to be rewritten to be more flexible

**Comment by otizonaizit** [comment]
I keep on saying this, but we need a sprint to start a deep refactoring of grader. The thing breaks at every commit, and it's painful when this happens when you are under pressure to get the reviews go or when you have to invite people.
I am not sure where we can find the time, but we really need a sprint. Anyone up for that? When?

**Comment by jni** [comment]
I would very much like to contribute to this effort, but as demonstrated by the pelita API overhaul, this is best done in-person. Perhaps @otizonaizit you want to come to ASPP-Apac and stay a bit to hack? ;)

As much as it pains me to say it, I think we should use pandas for the data structure. Its csv reader is very battle-tested and a lot of the work I did with fiddling with the applications.csv file with pandas was very easy.

**Comment by otizonaizit** [comment]
@jni : sure we could use pandas. But it's not like the standard library csv reader fails to read the `applications.csv` file. The problem of mapping fields in the csv file to fields in the named tuple, for past and future instances of the school, is where some of the pain in grader is. (But I have many other pain sources in grader). 

And I fail to see how a change to pandas would help us with that. At the end, we would still need to manually (or in the config) map the fields, and define equivalences within different spellings for the same fields (to keep on supporting old `applications.csv`). That is currently quite painful, and could, in my opinion, be done much more streamlined and easily. That is tangential to the pandas question. Regarding getting stats, I agree that pandas would give us stuff for free that we are currently doing manually. But at least that is not a place in grader that is causing **me** pain. 

Anyway, yes, an in-person sprint is definitely the way to go. Still, we could start with a video conference, like we did for the beginning of the pelita-API overhaul, and then, once the development directions have been more or less decided, we can continue with PRs and the standard OSS workflow. What do you think?

**Comment by jni** [comment]
@otizonaizit 

> And I fail to see how a change to pandas would help us with that.

True. It helps us with other things, at least two:
- filtering applications where some essential field is null; and probably handling applications uniformly when entire columns are missing from past years.
- computing the ranking, which currently takes ~30s-1min for 200 applications, which is frankly ridiculous. I was playing with the formula and regrading, and it was a very slow loop.

> Still, we could start with a video conference, like we did for the beginning of the pelita-API overhaul,

Sure. I just still feel bad that we never finished that effort and it fell on you and Rike to finally push it through, and I fear the same thing will happen here. Still, if you think it will help, I can do it. =)

**Comment by pberkes** [comment]
Last year when visiting @otizonaizit I had looked into using Pandas. It is feasible, but in the end I had given up because there are too many features that are considered to be necessary: e.g., each motivation score by each scorer must be available and editable in the configuration file, so there need to be a sync layer between the data structure and the file, so you need an object, so you end up losing the advantage of Pandas.

**Comment by pberkes** [comment]
As for the commit number dance, at a minimum we could add a tag for each school edition on both repositories.

**Comment by otizonaizit** [comment]
Sure, we could. For that, though, we'd need a release manager, or someone who feels responsible to (at least) do the tagging at the right moment. Finding the right moment is not 100% obvious: for example, I have the evaluation survey running after the school, but start getting mail of people who want to be put on the announcement list for the next edition right after the application deadline many months before the school even starts. But a bit of a more serious overhaul would be needed, for example by starting the work on the current edition directly in its subdirectory year-location, and not by moving it there at some point after the school. But for that, we need grader to know how to handle it.

**Comment by keszybz** [comment]
The `main` branch now has contains a refactoring that does similar or same changes. It's been so long that I don't even remember if those later changes are based on the ones here. But ultimately it doesn't matter, we need to keep with the work on the `main` branch until it does what is needed. This pull request is now obsolete.

File: grader/test_applications.py [hunk 1]
```diff
@@ -207,16 +241,67 @@ def test_applications_iteration():
...
+
+def test_person_motivation_scores():
+    scorers = ['me', 'you', 'him']
+    person_factory = build_person_factory(['name', 'lastname'],
+                                          scorers=scorers)
+    john_doe = person_factory('John', 'Doe')
+
+    john_doe.motivation_score = {'me': +1, 'you': -1, 'him': 0}
+    # use a set: we can't trust the order of a dictionary in Python < 3.6
+    assert set(john_doe.motivation_scores) == {1, -1, 0}
```
Comment (keszybz) [inline grader/test_applications.py:307]: This part looks useful. We should port it forward.

**Pull request closed without merging**